examples/udp_client/linux_udp_client.cc

```diff
@@ -0,0 +1,135 @@
+#include <cstring> // For memcpy()
+#include <string> // For stoi() and strerror()
+
+#include <netdb.h> // For gethostbyname() and hostent
+#include <netinet/in.h> // For IPPROTO_* macros and htons()
+#include <sys/socket.h> // For socket support.
+#include <unistd.h> // For close()
+
+#include <iostream>
+#include <fstream>
+#include <arpa/inet.h>
+
+#include <point_one/fusion_engine/parsers/fusion_engine_framer.h>
+
+#include "../common/print_message.h"
+
+#define DEBUG_ON 0
+
+using namespace point_one::fusion_engine::examples;
+using namespace point_one::fusion_engine::parsers;
+
+static bool shutdown_pending_ = false;
+
+/******************************************************************************/
+void HandleSignal(int signal) {
+  if (signal == SIGINT || signal == SIGTERM) {
+    std::signal(signal, SIG_DFL);
+    shutdown_pending_ = true;
+  }
+}
+
+void * get_in_addr(struct sockaddr * sa) {
+    if (sa->sa_family == AF_INET) {
+        return &(((struct sockaddr_in *)sa)->sin_addr);
+    }
+    return &(((struct sockaddr_in6 *)sa)->sin6_addr);
+}
+/******************************************************************************/
+
+int main(int argc, const char* argv[]) {
+  // Parse arguments.
+  if (argc > 2) {
+    printf(R"EOF(
+Usage: %s [PORT]
+
+Connect to an Atlas device over UDP and print out the incoming message
+contents.
+)EOF", 
+        argv[0]);
+    return 0;
+  }
+
+  int port = (argc == 2) ? std::stoi(argv[1]) : 12345;
+
+  // create UDP socket.
+  int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (sock < 0) {
+    printf("Error creating socket.\n");
+    return 2;
+  }
+
+  // bind socket to port
+  sockaddr_in addr;
+  addr.sin_family = AF_INET;
+  addr.sin_port = htons(port);
+  addr.sin_addr.s_addr = INADDR_ANY; // any local address
+  int ret = bind(sock, (struct sockaddr *) &addr, sizeof(addr));
+  if(ret < 0) {
+    close(sock);
+    printf("Error binding.\n");
+    return 3;
+  }
+
+  // Listen for SIGINT (Ctrl-C) or SIGTERM and shutdown gracefully.
+  std::signal(SIGINT, HandleSignal);
+  std::signal(SIGTERM, HandleSignal);
+
+  // Receive incoming data.
+  FusionEngineFramer framer(1024);
+  framer.SetMessageCallback(PrintMessage);
+
+  uint8_t buffer[1024];
+  size_t total_bytes_read = 0;
+  struct sockaddr_storage their_addr; // address of recieved packet
+  socklen_t addr_len = sizeof(their_addr);
+  ret = 0;
+  char their_ip[INET6_ADDRSTRLEN];
+  while (!shutdown_pending_) {
+    ssize_t bytes_read = recvfrom(sock, buffer, sizeof(buffer), 0,
+        (struct sockaddr *)&their_addr, &addr_len);
+
+    if (bytes_read < 0) {
+      printf("Error reading from socket: %s (%d)\n", std::strerror(errno),
+             errno);
+      ret = 4;
+      break;
+    } else if (bytes_read == 0) {
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+
+    inet_ntop(their_addr.ss_family, get_in_addr((struct sockaddr *)&their_addr), their_ip, sizeof(their_ip));
+    buffer[bytes_read] = '\0';
```

Comment:
This incoming data is almost always going to be binary, not a string, so printing it with `%s` below may do unexpected things and probably won't do anything too useful. I imagine this was here from your initial debugging? I'd suggest removing these lines and the if statement/print below now that this is all working. You can also then remove `their_addr` and `their_ip`, and use `recv()` instead of `recvfrom()`.

Reply:
As you mentioned before UDP, in contrast to TCP, is connectionless. Usually with TCP you have a connect() call to do the TCP 3-way handshake with the client and server but in this case the server is just spitting out packets with the ip address destined to the client machine. The client machine here just receives the packet on a specified port. That is really the only difference between TCP and UDP. In other words, `recvfrom()` is needed for UDP.

Reply:
Technically, I believe you can use either `recv()` or `recvfrom()` for UDP if you are not interested in the sender's IP address (`recvfrom()` accepts `nullptr` for the source arguments). From the `recv(2)` man page:

> Also, the following call
>   recv(sockfd, buf, len, flags);
> is equivalent to
>   recvfrom(sockfd, buf, len, flags, NULL, NULL);

In this case, it's probably safe to assume that the packet is from the Atlas device since for UDP you must first configure Atlas to send to your machine specifically, so we don't really need the `their_*` information. That being said, if you think it's useful to keep the debug print then it's probably ok.

In that case though, it's definitely wrap the call in a `#if DEBUG_ON` instead of `if (DEBUG_ON)` -- since `DEBUG_ON` is a constant, using an if statement will produce a `Statement always evaluates to false` warning on some compilers.

Reply:
I think that may just work. I just remember that I wanted to do just that in Vegas but for some reason I was having issues using recv() instead of recvfrom(). I would have used recv() but it was not working at the time. And I just remember the TCP client working with those system calls and noticing that the only difference was the connect() instead of bind() for UDP. After reading a little more I think it would be possible to use connect() for UDP transactions and then use recv() and send() bc the source and destination information will automatically be stored. Hmm I'd have to test it though. I wouldn't want to change something that I know works for a code style optimization that has not been tested. Ref: https://beej.us/guide/bgnet/html/index-wide.html#datagram, https://beej.us/guide/bgnet/html/index-wide.html#sendtorecv.

Reply:
great point about the #if preprocessor condition.